Clarify actor state store support for MongoDB (#3990)
dapr/docs: daprdocs/content/en/reference/components-reference/supported-state-stores/setup-mongodb.md

The above example uses secrets as plain strings. It is recommended to use a secret store for the secrets as described [here]({{< ref component-secrets.md >}}).
{{% /alert %}}

### Actor state store and transactions support

> For using as an actor state store or to leverage transactions, MongoDB must be running in a [Replica Set](https://www.mongodb.com/docs/manual/replication/).

Reviewer: ```suggestion
When using as an actor state store or to leverage transactions, MongoDB must be running in a [Replica Set](https://www.mongodb.com/docs/manual/replication/).
```

|--------------------|:--------:|---------|---------|
| server             | Y<sup>1</sup> | The server to connect to, when using DNS SRV record | `"server.example.com"`
| host               | Y<sup>1</sup> | The host to connect to | `"mongo-mongodb.default.svc.cluster.local:27017"`
| username           | N        | The username of the user to connect with (applicable in conjunction with `host`) | `"admin"`
| password           | N        | The password of the user (applicable in conjunction with `host`) | `"password"`
| databaseName       | N        | The name of the database to use. Defaults to `"daprStore"` | `"daprStore"`
| collectionName     | N        | The name of the collection to use. Defaults to `"daprCollection"` | `"daprCollection"`
| writeConcern       | N        | The write concern to use | `"majority"`
| readConcern        | N        | The read concern to use  | `"majority"`, `"local"`,`"available"`, `"linearizable"`, `"snapshot"`
| operationTimeout   | N        | The timeout for the operation. Defaults to `"5s"` | `"5s"`
| params             | N<sup>2</sup> | Additional parameters to use | `"?authSource=daprStore&ssl=true"`

> > <sup>[1]</sup> The `server` and `host` fields are mutually exclusive. If neither or both are set, Dapr will return an error.

Reviewer: ```suggestion
> <sup>[1]</sup> The `server` and `host` fields are mutually exclusive. If neither or both are set, Dapr returns an error.
```

> <sup>[2]</sup> The `params` field accepts a query string that specifies connection specific options as `<name>=<value>` pairs, separated by `&` and prefixed with `?`. e.g. to use "daprStore" db as authentication database and enabling SSL/TLS in connection, specify params as `?authSource=daprStore&ssl=true`. See [the mongodb manual](https://docs.mongodb.com/manual/reference/connection-string/#std-label-connections-connection-options) for the list of available options and their use cases.

> ## Setup MongoDB

Reviewer: It would be great if we could point to a resource on how to set up mongoDB replica sets easily as pointed out here https://stackoverflow.com/questions/51461952/mongodb-v4-0-transaction-mongoerror-transaction-numbers-are-only-allowed-on-a. Do you have a suggestion, or better still is there a public container image that could be downloaded and used?

Author: Done.

- Docker: I added a link to how to set up a RS using Docker. It's not a simple process, and requires a bunch of steps, so I think it's best to link to the official docs
- K8s: Looks like the Helm chart we recommended for K8s was archived. I've changed the recommendation to use the Helm chart by Bitnami, which also supports configuring Replica Sets. Since there are many options, and it depends on the architecture you choose (single node or RS), I've just added a link

| operationTimeout   | N        | The timeout for the operation. Defaults to `"5s"` | `"5s"`
| params             | N<sup>2</sup> | Additional parameters to use | `"?authSource=daprStore&ssl=true"`

> <sup>[1]</sup> The `server` and `host` fields are mutually exclusive. If neither or both are set, Dapr returns an error.

> <sup>[2]</sup> The `params` field accepts a query string that specifies connection specific options as `<name>=<value>` pairs, separated by `&` and prefixed with `?`. e.g. to use "daprStore" db as authentication database and enabling SSL/TLS in connection, specify params as `?authSource=daprStore&ssl=true`. See [the mongodb manual](https://docs.mongodb.com/manual/reference/connection-string/#std-label-connections-connection-options) for the list of available options and their use cases.

## Setup MongoDB

{{< tabs "Self-Hosted" "Kubernetes" >}}

{{% codetab %}}
You can run a single MongoDB instance locally using Docker:

```sh
docker run --name some-mongo -d mongo
```

You can then interact with the server at `localhost:27017`. If you do not specify a `databaseName` value in your component definition, make sure to create a database named `daprStore`.

> In order to use the MongoDB state store for transactions and as an actor state store, you will need to run MongoDB as a Replica Set. Please refer to [the official documentation](https://www.mongodb.com/compatibility/deploying-a-mongodb-cluster-with-docker) for how to create a 3-node Replica Set using Docker.

Reviewer: ```suggestion
In order to use the MongoDB state store for transactions and as an actor state store, you need to run MongoDB as a Replica Set. Refer to [the official documentation](https://www.mongodb.com/compatibility/deploying-a-mongodb-cluster-with-docker) for how to create a 3-node Replica Set using Docker.
```

{{% /codetab %}}

{{% codetab %}}

> You can conveniently install MongoDB on Kubernetes using the [Helm chart packaged by Bitnami](https://github.com/bitnami/charts/tree/main/bitnami/mongodb/). Refer to the documentation for the Helm chart for deploying MongoDB, both as a standalone server, and with a Replica Set (required for using transactions and actors).

Reviewer: ```suggestion
You can conveniently install MongoDB on Kubernetes using the [Helm chart packaged by Bitnami](https://github.com/bitnami/charts/tree/main/bitnami/mongodb/). Refer to the documentation for the Helm chart for deploying MongoDB, both as a standalone server, and with a Replica Set (required for using transactions and actors).
This installs MongoDB into the `default` namespace.
To interact with MongoDB, find the service with: `kubectl get svc mongo-mongodb`.
For example, if installing using the Helm defaults above, the MongoDB host address would be:
`mongo-mongodb.default.svc.cluster.local:27017`
Follow the on-screen instructions to get the root password for MongoDB.
The username is typically `admin` by default.
```

> You can conveniently install MongoDB on Kubernetes using the [Helm chart packaged by Bitnami](https://github.com/bitnami/charts/tree/main/bitnami/mongodb/). Please refer to the documentation for the Helm chart for deploying MongoDB, both as a standalone server, and with a Replica Set (required for using transactions and actors).

Reviewer: ```suggestion
You can conveniently install MongoDB on Kubernetes using the [Helm chart packaged by Bitnami](https://github.com/bitnami/charts/tree/main/bitnami/mongodb/). Refer to the documentation for the Helm chart for deploying MongoDB, both as a standalone server, and with a Replica Set (required for using transactions and actors).
```
